sled-agent/src/http_entrypoints.rs

```diff
@@ -64,9 +64,28 @@ async fn services_put(
     rqctx: RequestContext<SledAgent>,
     body: TypedBody<ServiceEnsureBody>,
 ) -> Result<HttpResponseUpdatedNoContent, HttpError> {
-    let sa = rqctx.context();
+    let sa = rqctx.context().clone();
     let body_args = body.into_inner();
-    sa.services_ensure(body_args).await.map_err(|e| Error::from(e))?;
+
+    // Spawn a task to run `services_ensure`: dropshot/hyper will cancel an
+    // endpoint's task if the call times out or is otherwise cancelled, and this
+    // leads to a specific issue where booting all the service zones takes
+    // longer than the progenitor client default timeout. The request times out,
+    // `services_ensure` gets cancelled, and this leaves zones partially
+    // configured but not added to the list of `existing_zones`. When the next
+    // `PUT /services` call is made, `services_ensure` will eventually try to
+    // bring up the same zone it was interrupted at, leading to configuration
+    // issues. See: oxidecomputer/omicron#3098.
+
+    match tokio::spawn(async move { sa.services_ensure(body_args).await }).await
+    {
+        Ok(result) => result.map_err(|e| Error::from(e))?,
+
+        Err(e) => {
+            return Err(HttpError::for_internal_error(e.to_string()));
```

Comment:
```suggestion
            return Err(HttpError::for_internal_error(&format!("failed to spawn tokio task for services_ensure: {:#}",e));
```

Reply:
I don't think the suggested message is right. I don't think `tokio::spawn()` itself is fallible: if it returns an error, it means the task that was spawned either panicked or was cancelled. I usually `.unwrap()` `tokio::spawn(..).await` calls that cannot be canceled (other than by the future calling them itself being canceled, which means you wouldn't get to the unwrap), since it can only panic if another panic has already happened.

Reply:
Good call.  I didn't read that closely enough.  In that case I'd suggest:

```rust
            return Err(HttpError::for_internal_error(&format!("unexpected failure awaiting \"services_ensure\" task: {:#}",e));
```

I'm not attached to the message.  I just don't like assuming that the message provided by the callee is going to be sufficient for someone looking through the logs to figure out what Nexus was trying to do when it ran into this problem.